File: src/main/scala/org/skycastle/entity/Action.scala
package org.skycastle.entity


import util.Parameters

/**
 * Some action available on an Entity.
 *
 * Can be forwarded to a local method call, or can be implemented as a Script.
 * 
 * @author Hans Haggstrom
 */
abstract case class Action(description : String) {

  def call( entity : Entity, parameters : Parameters )
}

case class MethodCallAction( description : String, methodName : String ) extends Action(description) {
  def call(entity: Entity, parameters: Parameters) {
    // Get method with specified name on entity, if found.
    // TODO

    // Call the method, pass the Parameters object as a parameter.
    // TODO

    println( "TODO: Implement calling " +methodName+" method on " +entity+ " with parameters " +parameters )
  }
}

case class ScriptCallAction( description : String, script : Script ) extends Action(description) {
  def call(entity: Entity, parameters: Parameters) {
    script.apply( entity, parameters )
  }
}

File: src/main/scala/org/skycastle/entity/action.java
package org.skycastle.entity;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

/**
 * Indicates that a method in an Entity is an action that can be called by another Entity through the action call mechanism.
 *
 * Specifies which incoming parameter names should be wrapped to which parameter, as a space separated string.
 *
 * @author Hans Haggstrom
 */
@Retention( RetentionPolicy.RUNTIME)
public @interface action
{
    /**
     * @return Space separated list of parameter names to use when calling the action method.
     */
    String value() default "";
}





